### PR TITLE
Implemented getsockname and poll

### LINKED ISSUE #2587: Implement getsockname and poll socket functions
### Describe the project you are working on
I'm currently working on a multiplayer game which uses STUN to get around the limitations of NAT.

### Describe the problem or limitation you are having in your project
`getsockname`: In the implementation I needed to be aware of what address the socket is bound to.
`poll`: I needed to set a timeout on receiving operations using an `EnetSocket`. So I need `poll`.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation
See above.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams
Probably obvious. Using platform's and Godot's socket functions.
Here is the pull request: godotengine/godot#41600

### If this enhancement will not be used often, can it be worked around with a few lines of script?
No it's can't be worked around because it's a core feature.

### Is there a reason why this should be core and not an add-on in the asset library?
Yes! Because it's a core feature. I know no other ways to do this.

### PR DESCRIPTION
**Pull request by MmAaXx500**
Implemented `getsocknme` in NetSocket and ENetSocket and `poll` in ENetSocket
This closes: godotengine/godot-proposals#2587

### CODE REVIEW

**Comment by Calinou** [comment]
Can you explain some possible use cases for this feature?

**Comment by MmAaXx500** [comment]
I used for STUN NAT behavior discovery in my multiplayer implementation.
I think it's useful if someone need similar multiplayer.

**Comment by aaronfranke** [comment]
@MmAaXx500 Is this still desired? If so, you should open a [proposal](https://github.com/godotengine/godot-proposals), and this needs to be [rebased](https://docs.godotengine.org/en/latest/community/contributing/pr_workflow.html#the-interactive-rebase) and tested on the latest master branch. While there are no conflicts, rebasing is important so that reviewers can easily test the PR. A proposal is also needed so that we can better evaluate what the use cases are for this feature.

**Comment by MmAaXx500** [comment]
I will put this into a draft until #48195 is resolved.

**Comment by MmAaXx500** [comment]
Updated the timeout handling according to #48203.
Now the Godot socket and Enet functions using the same unit, which is millisecond.

File: core/io/net_socket.h [hunk 1]
```diff
@@ -63,6 +63,7 @@ class NetSocket : public Reference {
 	virtual Error recvfrom(uint8_t *p_buffer, int p_len, int &r_read, IP_Address &r_ip, uint16_t &r_port, bool p_peek = false) = 0;
 	virtual Error send(const uint8_t *p_buffer, int p_len, int &r_sent) = 0;
 	virtual Error sendto(const uint8_t *p_buffer, int p_len, int &r_sent, IP_Address p_ip, uint16_t p_port) = 0;
+	virtual Error getsockname(IP_Address &r_ip, uint16_t &r_port) = 0;
```
[the remark this answers is not in the document]
Comment (MmAaXx500) [reply core/io/net_socket.h:66]: And in godot.cpp too?

[the remark this answers is not in the document]
Comment (aaronfranke) [reply core/io/net_socket.h:66]: Yes, this name would need to be changed in many places if it's changed on this line.

[the remark this answers is not in the document]
Comment (MmAaXx500) [reply core/io/net_socket.h:66]: The godot.cpp is a separate class and is in the thirdparty folder, that's why I asked. 
Ok, I'll modify it tomorrow (it's late at night here).

**Comment by Faless** [comment]
See #48235 for a `get_socket_address` implementation in ENet.
And #47398 for the `NetSocket` one.

**Comment by MmAaXx500** [comment]
Then this pr is useless, right?

**Comment by Faless** [comment]
> Then this pr is useless, right?

I think the only unimplemented bit there is `enet_socket_wait`.

**Comment by MmAaXx500** [comment]
> I think the only unimplemented bit there is `enet_socket_wait`.

Is it desirable? If so, I'll do my best in a separate pr.
BTW thanks for the better implementation of getsockname.

BTW2: Some hours ago I discovered the fact that I'm currently not using the `wait`. :laughing:  Probably implemented when I planned the things for my game.

**Pull request closed without merging**